### ISSUE TITLE
FB9857161: NSAlert crashes app if called from a Task

### ISSUE THREAD

**Issue by sindresorhus** [feedback-assistant/reports#288]
- **Date:** 2022-01-24
- **Resolution:** Open
- **Area:** AppKit
- **OS:** macOS 12.1
- **Type:** Application Crash

## Description

NSAlert crashes when called from a Task, even if you use `@MainActor`. This is a pretty serious issue as a lot of code uses NSAlert.

I have attached a sample project that reproduces the issue.

```swift
Task { @MainActor in
	let alert = NSAlert()
	alert.messageText = "X"
	alert.runModal()
}
```

Stack trace:

```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x00000001c7c37a18 in libunwind::UnwindCursor<libunwind::LocalAddressSpace, libunwind::Registers_arm64>::step() ()
#1	0x00000001bd0b762c in objc_addExceptionHandler ()
#2	0x00000001c00e5e80 in -[NSApplication _commonBeginModalSessionForWindow:relativeToWindow:modalDelegate:didEndSelector:contextInfo:] ()
#3	0x00000001c00e5a50 in __35-[NSApplication runModalForWindow:]_block_invoke_2 ()
#4	0x00000001c00e5a08 in __35-[NSApplication runModalForWindow:]_block_invoke ()
#5	0x00000001c00e5198 in _NSTryRunModal ()
#6	0x00000001c00e5048 in -[NSApplication runModalForWindow:] ()
#7	0x00000001c024e218 in __19-[NSAlert runModal]_block_invoke_2 ()
#8	0x00000001c024e15c in __19-[NSAlert runModal]_block_invoke ()
#9	0x00000001c00e5198 in _NSTryRunModal ()
#10	0x00000001c016f694 in -[NSAlert runModal] ()
#11	0x0000000104b69ddc in closure #1 in ViewController.viewDidLoad()
```

## Files

[Sample project.zip](https://github.com/feedback-assistant/reports/files/7925521/Sample.project.zip)

**Comment by sindresorhus** [comment]
Hacky workaround:

```swift
extension NSAlert {
	/**
	Workaround to allow using `NSAlert` in a `Task`.

	[FB9857161](https://github.com/feedback-assistant/reports/issues/288)
	*/
	@MainActor
	@discardableResult
	func run() async -> NSApplication.ModalResponse {
		await withCheckedContinuation { continuation in
			DispatchQueue.main.async { [self] in
				continuation.resume(returning: runModal())
			}
		}
	}
}
```

**Comment by Kyle-Ye** [comment]
+1. (Encountered the issue when refactoring to swift-concurrency)

Neither `MainActor.run { alert.runModal() }`, `Task { await alert.runModal() }` and `Task { @MainActor alert.runModal()}` works for me. The same crash and stack trace.
Thanks for your workaround.

<img width="1552" alt="截屏2022-02-15 13 18 32" src="https://user-images.githubusercontent.com/43724855/153997694-018aa97b-5263-4a88-bc55-acd398d6fcea.png">

**Comment by Kyle-Ye** [comment]
By the way, in addition to `NSAlert`, `NSOpenPanel` and other `NSView` containing `runModal` methods will also crash for the same reason

**Comment by sindresorhus** [comment]
For `NSOpenPanel`, you can just use `await panel.begin()` (the auto-transformed completion handler method).

**Comment by sindresorhus** [comment]
Also, don't forget to duplicate my report, to make it more likely Apple will see it and listen.

**Comment by krzyzanowskim** [comment]
is seems to be fixed when run in macOS 13 (beta5). Still crashes when run on macOS 12

**Comment by sindresorhus** [comment]
Fixed in macOS 13.3.1
